### PR TITLE
fix freezing with deepcopy model

### LINKED ISSUE #2905: Deepcopy of model with categorical features causes infinite hang during prediction after refitting
**Problem**
Модель виснет на этапе инференса. 

**Описание**
В  коде копируем исходную модель, переобучаем  и делаем предсказание, на котором часто зависает.  
Без категориальных признаков отрабатывает всегда. Ниже привожу простой пример,  на котором зависает выполнение. Есть подозрение на строчку  deepcopy(model), и  .drop(features_to_drop,  axis=1). 

```python
import pandas as pd
import numpy as np
from catboost import CatBoostClassifier, Pool
from sklearn.model_selection import train_test_split
from copy import deepcopy

np.random.seed(0)

n_samples = 40*10**3

num_features = pd.DataFrame({
    'num_1': np.random.randn(n_samples),
    'num_2': np.random.rand(n_samples) * 100,
    'num_3': np.random.randint(0, 50, n_samples)
})


cat_features = pd.DataFrame({
    'cat_1': np.random.choice(['A', 'B', 'C'], n_samples),
    'cat_2': np.random.choice(['X', 'Y'], n_samples),
    'cat_3': np.random.choice(['red', 'green', 'blue', 'yellow'], n_samples)
})


target = ((num_features['num_1'] + 
           (cat_features['cat_1'] == 'B').astype(int) - 
           (cat_features['cat_3'] == 'green').astype(int) + 
           np.random.randn(n_samples)*0.5) > 0).astype(int)

df = pd.concat([num_features, cat_features], axis=1)

df_train, df_test, target_train, target_test = train_test_split(df, target, test_size=0.3, random_state=42)

cat_features_indices = [df_train.columns.get_loc(col) for col in ['cat_1', 'cat_2', 'cat_3']]

model = CatBoostClassifier(
    iterations=40,
    learning_rate=0.25,
    depth=11,
    verbose=False,
    random_seed=42, 
    class_weights=[1, 20],
    thread_count=3, 
    l2_leaf_reg=50
)

model.fit(
    df_train, target_train,
    cat_features=cat_features_indices
)

features_to_drop = []
def refit_and_predict(model, df_train, target_train, df_test, features_to_drop, cat_features_indices):
    from copy import deepcopy

    model_copy = deepcopy(model)

    model_copy = model_copy.fit(
        df_train.drop(features_to_drop, axis=1),
        target_train,
        cat_features=cat_features_indices   
    )
    

    # **problem here**
    refit_preds_train = model_copy.predict(df_train.drop(features_to_drop, axis=1))
    refit_preds_test = model_copy.predict(df_test.drop(features_to_drop, axis=1))

    return refit_preds_train, refit_preds_test
    

train_preds, test_preds = refit_and_predict(model, df_train, target_train, df_test, features_to_drop, cat_features_indices)
```
###requirements
pandas==1.4.3
numpy==1.26.4
catboost==1.2.8
scikit-learn==1.5.2
ipykernel==6.29.5

version of python  3.9
os: SberLinux

### PR DESCRIPTION
**Pull request by SergeevVladislav**
Fixes https://github.com/catboost/catboost/issues/2905

### CODE REVIEW

File: catboost/python-package/catboost/_catboost.pyx [hunk 1]
```diff
@@ -5830,8 +5831,12 @@ cdef class _CatBoost:
         return py_serialized_model_str
 
     cpdef _deserialize_model(self, serialized_model_str):
-        self.model_blob = serialized_model_str
-        cdef TFullModel tmp_model = ReadZeroCopyModel(<char*>serialized_model_str, len(serialized_model_str))
+        # Use ReadModel instead of ReadZeroCopyModel to avoid use-after-free
+        # when model_blob is cleared during subsequent fit() calls.
+        # See https://github.com/catboost/catboost/issues/2905
+        cdef EModelType modelType = string_to_model_type('cbm')
+        cdef TFullModel tmp_model = ReadModel(<char*>serialized_model_str, len(serialized_model_str), modelType)
```
Comment (andrey-khropov) [inline catboost/python-package/catboost/_catboost.pyx]: Replacing `ReadZeroCopyModel` with `ReadModel` is ineffective - additional data structures that aren't needed when we use `ReadZeroCopyModel` will be allocated and filled, it means an additional copy of the model data will consume RAM, even if only temporarily.

I suggest to fix the logic around `model_blob` instead. It must always be cleared only after the corresponding `TFullModel` object in `self.__model` is destroyed.

**Review by andrey-khropov — CHANGES_REQUESTED** [review]

**Review by SergeevVladislav — COMMENTED** [review]
@andrey-khropov 
ReadZeroCopyModel has been saved (eliminating the extra copy), and the logic has been restructured around the lifetime of `model_blob` as requested: the blob is now always freed only after the `TFullModel` referencing it is destroyed.

**Review by andrey-khropov — CHANGES_REQUESTED** [review]
Let's avoid copy-paste 

```
    cdef TFullModel* new_model = new TFullModel()
    new_model[0].Swap(tmp_model)
    self._replace_model(new_model, None)
```


and make another member function in class `_CatBoost` like

```
cdef _replace_model_ref(self, TFullModel& new_model_ref, object new_model_blob):
     cdef TFullModel* new_model = new TFullModel()
     new_model[0].Swap(new_model_ref)		
     self._replace_model(new_model, new_model_blob)
```

and use it instead.

**Review by andrey-khropov — APPROVED** [review]

**Comment by andrey-khropov** [comment]
LGTM

**Comment by andrey-khropov** [comment]
Tests failed in CI because of the issue with using unstable `plotly` version in dependencies. It has been fixed in https://github.com/catboost/catboost/commit/e10520bb8ce6b13b1264baa902fef2c021697232
